File: sandbox/src/layer2_clamav.c
/* Layer 2: ClamAV integration via secure subprocess execution.
 *
 * Spawns clamdscan or clamscan as a child process using the Win32
 * CreateProcess API — no shell, no system(), immune to injection.
 * Parses output to extract virus names.
 */
#include "layer2_clamav.h"
#include "subprocess.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define CLAMAV_TIMEOUT 30

/* Find first available ClamAV binary. Returns name or NULL. */
static const char *find_clamav(void) {
    static const char *bins[] = {"clamdscan.exe", "clamscan.exe",
                                 "clamdscan", "clamscan", NULL};
    int i;
    for (i = 0; bins[i]; i++) {
        if (subprocess_which(bins[i]))
            return bins[i];
    }
    return NULL;
}

/* Parse virus name from ClamAV output line.
 * Format: "/path/to/file: Virus.Name.Here FOUND"
 */
static void parse_virus_name(const char *output, char *name, size_t name_size) {
    const char *colon = strchr(output, ':');
    const char *found;
    size_t len;

    if (!colon) {
        snprintf(name, name_size, "%s", output);
        return;
    }

    colon++; /* skip colon */
    while (*colon == ' ') colon++;

    found = strstr(colon, "FOUND");
    if (found) {
        len = (size_t)(found - colon);
        while (len > 0 && colon[len-1] == ' ') len--;
        if (len >= name_size) len = name_size - 1;
        memmove(name, colon, len);
        name[len] = '\0';
    } else {
        snprintf(name, name_size, "%s", colon);
    }
}

Layer2Result layer2_scan(const char *file_path) {
    Layer2Result r;
    const char *clamav_bin;
    char *argv[8];
    char stdout_buf[8192];
    char stderr_buf[4096];
    int exit_code, spawn_ret;

    memset(&r, 0, sizeof(r));
    r.status = 2; /* default error */

    clamav_bin = find_clamav();
    if (!clamav_bin) {
        r.status = 2;
        snprintf(r.reason, sizeof(r.reason),
                 "ClamAV is not installed or not reachable on this system");
        return r;
    }

    /* Build argv: clamscan --no-summary --stdout <file> */
    argv[0] = (char *)clamav_bin;
    argv[1] = (char *)"--no-summary";
    argv[2] = (char *)"--stdout";
    argv[3] = (char *)file_path;
    argv[4] = NULL;

    spawn_ret = subprocess_run(
        clamav_bin, argv, CLAMAV_TIMEOUT,
        &exit_code,
        stdout_buf, sizeof(stdout_buf),
        stderr_buf, sizeof(stderr_buf)
    );

    if (spawn_ret == -1) {
        r.status = 2;
        snprintf(r.reason, sizeof(r.reason), "Failed to launch ClamAV scanner");
        return r;
    }
    if (spawn_ret == -2) {
        r.status = 2;
        snprintf(r.reason, sizeof(r.reason),
                 "ClamAV scan timed out after %d seconds", CLAMAV_TIMEOUT);
        return r;
    }

    /* clamscan exit codes: 0=clean, 1=virus found, 2=error */
    if (exit_code == 0) {
        r.status = 0; /* clean */
        return r;
    }

    if (exit_code == 1) {
        const char *output = stdout_buf[0] ? stdout_buf : stderr_buf;
        r.status = 1; /* infected */
        parse_virus_name(output, r.virus_name, sizeof(r.virus_name));
        if (r.virus_name[0]) {
            snprintf(r.reason, sizeof(r.reason),
                     "Malware detected: %s", r.virus_name);
        } else {
            snprintf(r.reason, sizeof(r.reason), "Malware detected: Unknown threat");
        }
        return r;
    }

    /* exit_code == 2 or other: error */
    r.status = 2;
    if (stderr_buf[0]) {
        snprintf(r.reason, sizeof(r.reason),
                 "ClamAV scan error (exit %d): %s", exit_code, stderr_buf);
    } else {
        snprintf(r.reason, sizeof(r.reason),
                 "ClamAV scan error (exit code %d)", exit_code);
    }
    return r;
}

File: sandbox/layers/layer2_clamav.py
"""Layer 2: Static Antivirus Scan via ClamAV.

Interfaces with a locally installed ClamAV daemon (clamdscan) or standalone
scanner (clamscan). Gracefully handles the case where ClamAV is unavailable.
"""

import json
import os
import subprocess
import sys
import tempfile
from typing import Optional

from config import CLAMAV_COMMANDS, CLAMAV_TIMEOUT_SECONDS


class Layer2Result:
    __slots__ = ("status", "reason", "virus_name", "raw_output")

    def __init__(self, status: str, reason: str = "", virus_name: str = "", raw_output: str = ""):
        self.status = status
        self.reason = reason
        self.virus_name = virus_name
        self.raw_output = raw_output

    def to_dict(self) -> dict:
        d = {"status": self.status}
        if self.reason:
            d["reason"] = self.reason
        if self.virus_name:
            d["virus_name"] = self.virus_name
        if os.environ.get("SANDBOX_DEBUG"):
            d["clamav_raw"] = self.raw_output
        return d


def _find_clamav() -> Optional[str]:
    """Locate an available ClamAV binary on the system PATH."""
    for cmd in CLAMAV_COMMANDS:
        try:
            result = subprocess.run(
                [cmd, "--version"],
                capture_output=True,
                text=True,
                timeout=5,
            )
            if result.returncode == 0:
                return cmd
        except (FileNotFoundError, subprocess.TimeoutExpired):
            continue
    return None


def scan(file_path: str) -> Layer2Result:
    """Scan a file using ClamAV.

    Returns Layer2Result with status 'clean', 'infected', or 'error'.
    """
    clamav_bin = _find_clamav()

    if not clamav_bin:
        return Layer2Result(
            status="error",
            reason="Layer 2: ClamAV is not installed or not reachable on this system",
        )

    abs_path = os.path.abspath(file_path)

    try:
        proc = subprocess.run(
            [clamav_bin, "--no-summary", "--stdout", abs_path],
            capture_output=True,
            text=True,
            timeout=CLAMAV_TIMEOUT_SECONDS,
        )

        stdout = proc.stdout.strip()
        stderr = proc.stderr.strip()

        # clamdscan/clamscan exit codes:
        #   0 = clean
        #   1 = virus found
        #   2 = error
        if proc.returncode == 0:
            return Layer2Result(
                status="clean",
                raw_output=stdout if stdout else stderr,
            )

        if proc.returncode == 1:
            # Virus found — extract the virus name from output
            virus_name = _parse_virus_name(stdout)
            return Layer2Result(
                status="infected",
                reason="Layer 2: Malware detected — {}".format(virus_name or "Unknown threat"),
                virus_name=virus_name or "Unknown.Threat",
                raw_output=stdout,
            )

        # returncode == 2 or other: error
        return Layer2Result(
            status="error",
            reason="Layer 2: ClamAV scan error (exit code {}): {}".format(
                proc.returncode, stderr or stdout or "unknown error"
            ),
            raw_output=stderr,
        )

    except subprocess.TimeoutExpired:
        return Layer2Result(
            status="error",
            reason="Layer 2: ClamAV scan timed out after {} seconds".format(CLAMAV_TIMEOUT_SECONDS),
        )
    except Exception as exc:
        return Layer2Result(
            status="error",
            reason="Layer 2: Unexpected error during ClamAV scan: {}".format(exc),
        )


def _parse_virus_name(output: str) -> str:
    """Extract the virus/malware name from ClamAV output.

    Typical format: /path/to/file: Virus.Name.Here FOUND
    """
    if ":" in output:
        after_path = output.split(":", 1)[1].strip()
        if "FOUND" in after_path:
            return after_path.replace("FOUND", "").strip()
        return after_path
    return output.strip()


def main():
    """Standalone CLI entry for Layer 2."""
    if len(sys.argv) != 2:
        print(json.dumps({"status": "error", "reason": "Usage: layer2_clamav.py <file_path>"}))
        sys.exit(1)
    result = scan(sys.argv[1])
    print(json.dumps(result.to_dict()))
    sys.exit(0 if result.status == "clean" else 1)


if __name__ == "__main__":
    main()
